File: src/main/kotlin/org/kite9/diagram/common/algorithms/ssp/State.java
package org.kite9.diagram.common.algorithms.ssp;

import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

public class State<P extends PathLocation<P>> {

	protected PriorityQueue<P> pq = new PriorityQueue<P>();
	private Map<Object, P> locationToPathMap = new HashMap<Object, P>(2000);
	public long adds;
	public long maxStack;
	private AbstractSSP<P> ssp;
	
	public State(AbstractSSP<P> in) {
		this.ssp = in;
	}

	public boolean add(P path) {
		try {
			adds ++;
			Object location = ssp.getLocation(path);
			P existing = location == null ? null : locationToPathMap.get(location);
			boolean newBetter = existing == null || existing.compareTo(path) > 0;

			if (newBetter) {
				if (existing != null) {
					existing.setActive(false);
				}
				//System.out.println("Replacing: \n\t "+existing+"\n\t"+path+"\n\t"+location);
				pq.add(path);
				locationToPathMap.put(location, path);
				maxStack = Math.max(pq.size(), maxStack);
				return true;
			} else {
				//System.out.println("Not Adding: "+path);
				return false;
			}
		} catch (OutOfMemoryError e) {
			throw new SSPTooLargeException("SSP too large: queue="+pq.size()+" map="+locationToPathMap.size());
		}
	}
	
	public P remove() {
		return pq.remove();
	}
}

File: kite9-visualization/src/commonMain/kotlin/org/kite9/diagram/common/algorithms/ssp/State.kt
package org.kite9.diagram.common.algorithms.ssp

open class State<P : PathLocation<P>>(private val ssp: AbstractSSP<P>) {

	var pq = PriorityQueue<P>(10000)
    private val locationToPathMap: MutableMap<Any?, P> = HashMap(2000)
	private var adds: Long = 0
	private var maxStack: Long = 0

    fun add(path: P): Boolean {
        return try {
            adds++
            val location = ssp.getLocation(path)
            val existing =  locationToPathMap[location]
            val newBetter = existing == null || existing.compareTo(path) > 0
            if (newBetter) {
                existing?.setActive(false)
                //System.out.println("Replacing: \n\t "+existing+"\n\t"+path+"\n\t"+location);
                pq.add(path)
                locationToPathMap[location] = path
                maxStack = pq.size().toLong().coerceAtLeast(maxStack)
                true
            } else {
                //System.out.println("Not Adding: "+path);
                false
            }
        } catch (e: Exception) {
            throw SSPTooLargeException("SSP too large: queue=" + pq.size() + " map=" + locationToPathMap.size)
        }
    }

    open fun remove(): P {
        return pq.remove()!!
    }

    fun getAdds() : Long {
        return adds;
    }

    fun getMaxStack() : Long {
        return maxStack;
    }
 }